### PR TITLE
kube-proxy: Fix flag validation for healthz-bind-address and metrics-bind-address

### LINKED ISSUE #53754: Kube-proxy metrics bind
> /kind bug

Whenever I try to change kube proxy metrics bind using flag `--metrics-bind-address=0.0.0.0`, it gives me:
```
Oct 12 07:25:50 R11U2-7U20N-1 kube-proxy[510997]: E1012 07:25:50.358481  510997 server.go:661] starting metrics server failed: listen tcp: address 0.0.0.0: missing port in address
```
Then, I added the port `--metrics-bind-address=0.0.0.0:10249`, the error changes.
```
Oct 12 07:25:58 R11U2-7U20N-1 kube-proxy[510947]: Error: invalid argument "127.0.0.1:10249" for --metrics-bind-address=0.0.0.0:10249: "0.0.0.0:10249" is not a valid IP address
```

I expect the metrics server to listen to 0.0.0.0:10249. Do I miss something?

**Environment**:
- Kubernetes version: v1.7.8
- Cloud provider or hardware configuration**: on-premise
- OS (e.g. from /etc/os-release): debian/stretch
- Kernel (e.g. `uname -a`): Linux R11U2-7U20N-1 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u3 (2017-08-06) x86_64 GNU/Linux
- Install tools: 
- Others:

### PR DESCRIPTION
**Pull request by MrHohn**
**What this PR does / why we need it**: `--healthz-bind-address` and `--metrics-bind-address` are broken for kube-proxy as they do not allow `ip:port` format, though they claim to support it.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: Fixes #53754

**Special notes for your reviewer**:
cc @ncdc 

**Release note**:

```release-note
Fix kube-proxy flags validation for --healthz-bind-address and --metrics-bind-address to allow specifying ip:port.
```

### CODE REVIEW

**Comment by ncdc** [comment]
@MrHohn do you plan to rebase this (to match the move from componentconfig to pkg/proxy/apis/kubeproxyconfig) for 1.9?

**Comment by MrHohn** [comment]
@ncdc Thanks for the reminder, I missed the rebase-needed signal. Will rebase and let you know.

**Comment by MrHohn** [comment]
@ncdc Rebased. Though `IPPortVar` is kept in `componentconfig/helpers.go` as I haven't figured out a better place.

**Comment by MrHohn** [comment]
Thanks for the quick review.

/assign @mikedanese 
for approval.

**Comment by BenTheElder** [comment]
/test pull-kubernetes-cross
[borrowing legitimate PR for test-infra purposes :-)]

**Comment by MrHohn** [comment]
@bowei @thockin It might be worth to get this in 1.9 (and cherrypick to 1.8, 1.7 as well). These two flags (especially `--metrics-bind-address`) are almost unusable without the fix.

File: pkg/apis/componentconfig/helpers.go [hunk 1]
```diff
@@ -61,6 +63,41 @@ func (v IPVar) Type() string {
 	return "ip"
 }
 
+type IPPortVar struct {
+	Val *string
+}
+
+func (v IPPortVar) Set(s string) error {
+	ip := s
+	if index := strings.Index(s, ":"); index != -1 {
```
Comment (bowei) [inline pkg/apis/componentconfig/helpers.go]: how does this handle ipv6?

`[aa:bb:cc::00]:80`

Comment (MrHohn) [reply pkg/apis/componentconfig/helpers.go]: Good catch, I sent this out without considering ipv6. Will fix.

Comment (thockin) [reply pkg/apis/componentconfig/helpers.go]: net.SplitHostPort()

Comment (MrHohn) [reply pkg/apis/componentconfig/helpers.go]: @thockin Thanks, that's something I've been looking for.

**Comment by MrHohn** [comment]
Updated PR for ipv6.

File: pkg/apis/componentconfig/helpers.go [hunk 2]
```diff
@@ -61,6 +62,49 @@ func (v IPVar) Type() string {
...
+	Val *string
+}
+
+func (v IPPortVar) Set(s string) error {
+	if len(s) == 0 {
+		v.Val = nil
+		return nil
+	}
+	// Both IP and IP:port are valid.
+	if net.ParseIP(s) == nil {
```
Comment (bowei) [inline pkg/apis/componentconfig/helpers.go]: reads better if you just bail out early
```
if net.ParseIP(s) != nil {
  *v.Val = s
  return nil
}

...
```

Comment (MrHohn) [reply pkg/apis/componentconfig/helpers.go]: Sounds good, done.

**Comment by MrHohn** [comment]
@bowei @thockin PR is updated. Thanks!

**Pull request merged**